conans/client/build/cmake.py

```diff
@@ -175,10 +175,15 @@ def _run(self, command):
         the_os = self._settings.get_safe("os")
         is_clangcl = the_os == "Windows" and compiler == "clang"
         is_msvc = compiler == "Visual Studio"
-        if (is_msvc or is_clangcl) and self.generator in ["Ninja", "NMake Makefiles",
-                                                          "NMake Makefiles JOM"]:
-            with tools.vcvars(self._settings, force=True, filter_known_paths=False,
-                              output=self._conanfile.output):
+        if ((is_msvc or is_clangcl) and platform.system() == "Windows" and
+                self.generator in ["Ninja", "NMake Makefiles", "NMake Makefiles JOM"]):
+            vcvars_dict = tools.vcvars_dict(self._settings, force=True, filter_known_paths=False,
+                                            output=self._conanfile.output)
+            if vcvars_dict:
+                with environment_append(vcvars_dict):
+                    with environment_append(self._conanfile.env):
+                        self._conanfile.run(command)
+            else:
```

Comment:
so, the problem was missing `self._conanfile.env` in environment?
if so, why don't we apply it for the `else` blocks as well?

Reply:
The `conanfile.env` is applied before running `build()` function in conanfile...

I also find this if/else ugly, I would prefer to apply the `conanfile.env` twice and clean this bit (it should have no side effects)... but in the end we would need to refactor this function and probably pass an argument `env` to the functions, add to it, and apply it just before running the command.